Src/Base/AMReX_IntegratorBase.H

```diff
@@ -0,0 +1,191 @@
+    {
+        // Emplace a new T in V with the same size as Other
+        V.emplace_back(std::make_unique<T>());
+        for (auto const& other_mf : Other) {
+            int nGrow = Grow ? other_mf->nGrow() : 0;
+            V.back()->emplace_back(std::make_unique<amrex::MultiFab>(other_mf->boxArray(), other_mf->DistributionMap(), other_mf->nComp(), nGrow));
+        }
+    }
+
+    static void Copy(T& Y, const T& Other, bool Grow = true)
+    {
+        // Copy the contents of Other into Y
+        const int size = Y.size();
+        for (int i = 0; i < size; ++i) {
+            int nGrow = Grow ? Other[i]->nGrow() : 0;
+            amrex::MultiFab::Copy(*Y[i], *Other[i], 0, 0, Other[i]->nComp(), nGrow);
+        }
+    }
+
+    static void Saxpy(T& Y, const amrex::Real a, const T& X, bool Grow = false)
+    {
+        // Calculate Y += a * X
+        const int size = Y.size();
+        for (int i = 0; i < size; ++i) {
+            int nGrow = Grow ? X[i]->nGrow() : 0;
+            amrex::MultiFab::Saxpy(*Y[i], a, *X[i], 0, 0, X[i]->nComp(), nGrow);
+        }
+    }
+
+};
+
+template<class T>
+struct IntegratorOps<T, typename std::enable_if<std::is_same<amrex::MultiFab, T>::value>::type>
+{
+
+    static void CreateLike(amrex::Vector<std::unique_ptr<T> >& V, const T& Other, bool Grow = false)
+    {
+        // Emplace a new T in V with the same size as Other
+        int nGrow = Grow ? Other.nGrow() : 0;
+        V.emplace_back(std::make_unique<T>(Other.boxArray(), Other.DistributionMap(), Other.nComp(), nGrow));
+    }
+
+    static void Copy(T& Y, const T& Other, bool Grow = true)
+    {
+        // Copy the contents of Other into Y
+        int nGrow = Grow ? Other.nGrow() : 0;
+        amrex::MultiFab::Copy(Y, Other, 0, 0, Other.nComp(), nGrow);
+    }
+
+    static void Saxpy(T& Y, const amrex::Real a, const T& X, bool Grow = false)
+    {
+        // Calculate Y += a * X
+        int nGrow = Grow ? X.nGrow() : 0;
+        amrex::MultiFab::Saxpy(Y, a, X, 0, 0, X.nComp(), nGrow);
+    }
+
+};
+
+template<class T>
+class IntegratorBase
+{
+private:
+    std::function<void(T&, const T&, const amrex::Real)> Fun;
+
+protected:
+    std::function<void (T&, amrex::Real)> post_update;
+
+public:
+    amrex::Real timestep;
+    IntegratorBase() {}
+
+    virtual ~IntegratorBase() {}
+
+    void set_rhs(std::function<void(T&, const T&, const amrex::Real)> F)
+    {
+        Fun = F;
+    }
+
+    void set_post_update(std::function<void (T&, amrex::Real)> F)
+    {
+        post_update = F;
+    }
+
+    virtual amrex::Real advance(T& S_old, T& S_new, amrex::Real time, const amrex::Real dt)
+    {
+        return dt;
+    }
+
+    void rhs(T& S_rhs, const T& S_data, const amrex::Real time)
+    {
+        Fun(S_rhs, S_data, time);
+    }
+
+    virtual void time_interpolate(const T& S_new, const T& S_old, amrex::Real timestep_fraction, T& data) {}
+
+    virtual void map_data(std::function<void(T&)> Map) {}
```

Comment:
Does it make sense to make these two virtual functions pure virtual?

Reply:
yes, thanks, I'm making these and advance() pure virtual